nservicebus/pipeline/customizing-v6.md

```diff
@@ -0,0 +1,75 @@
+---
+title: Customizing the Pipeline in NSB v6
+summary: Customizing the message handling pipeline in NServiceBus version 6.
+tags:
+- Pipeline
+related:
+- nservicebus/pipeline/customizing-v5
+- nservicebus/pipeline/customizing-v6
+---
+
+NServiceBus version 6 splits the existing pipelines into smaller and better composable units called "Stages". A stage is a group of steps acting on the same level of abstraction. This allows you to define a step working with the "incoming physical" message before it has been deserialized or a step which is executed before and after each handler invocation (remember: there can be multiple message handlers per message). You'll find more information about the existing stages later in this article.
+
+You can extend the pipeline with a custom behavior implementing `Behavior<TContext>`. `TContext` is the context of the stage that the behavior is working in.
+
+snippet:SamplePipelineBehavior
+
+In the above code snippet the `SampleBehavior` class derives from `Behavior<LogicalMessageProcessingContext>`. This tells the framework to execute this behavior after the incoming raw message has been deserialized and a matching message type has been found. At runtime, the pipeline will call the `Invoke` method of each registered behavior passing in as arguments the current message context and an action to invoke the next behavior in the pipeline.
+
+Warning: Each behavior is responsible to call the next step in the pipeline chain by invoking `next`.
+
+
+## Add a new step
+
+To add your custom behavior to the pipeline, you need to define a step for it:
+
+snippet:NewPipelineStep
+
+Then register the new step in the pipeline settings:
+
+snippet:AddPipelineStep
+
+
+## Replace an existing step
+
+To replace the implementation of an existing step, you can replace it with your custom behavior:
+
+snippet:ReplacePipelineStep
+
+Note: You can also register your steps from a [Feature](features.md)
+
+
+## Stages
```

Comment:
@johnsimons I realized that option too, here are my thoughts:
* Since we're now having HandlerContext and BehaviorContexts, this is quite a "context soup" where I doubt an end user will really understand the difference. So using *Stage instead of *Context would probably help.
* The BehaviorContexts are still the context of that stage, not directly the stage itself I guess. So that may be confusing too?

I'd like to discuss your idea and see if we can come up with a good solution from a user perspective :+1: